gdal/ogr/ogrsf_frmts/flatgeobuf/ogrflatgeobuflayer.cpp

```diff
@@ -0,0 +1,829 @@
+        CPLDebug("FlatGeobuf", "Wrote feature buffers (%zu bytes)", c);
+        offset += c;
+
+        VSIFCloseL(fp);
+    }
+
+    if (m_poFeatureDefn != nullptr)
+        m_poFeatureDefn->Release();
+
+    if (m_poSRS)
+        m_poSRS->Release();
+
+    if (m_featureBuf != nullptr)
+        VSIFree(m_featureBuf);
+
+    if (m_featureOffsets != nullptr)
+        VSIFree(m_featureOffsets);
+
+    m_processedSpatialIndex = false;
+}
+
+OGRFeature *OGRFlatGeobufLayer::GetFeature(GIntBig /* nFeatureId */)
+{
+    CPLError(CE_Fatal, CPLE_AppDefined, "GetFeature: Not implemented");
+    return nullptr;
+}
+
+void OGRFlatGeobufLayer::processSpatialIndex() {
+    if (m_poFilterGeom != nullptr && !m_processedSpatialIndex) {
+        CPLDebug("FlatGeobuf", "Attempting spatial index search");
+        if (m_poHeader->index_node_size() == 0) {
+            CPLError(CE_Warning, CPLE_AppDefined, "No spatial index available");
+            return;
+        }
+        m_processedSpatialIndex = true;
+        if (m_poFp == nullptr) {
+            CPLDebug("FlatGeobuf", "processSpatialIndex (will attempt to open file %s)", m_pszFilename);
+            m_poFp = VSIFOpenL(m_pszFilename, "rb");
+            //m_poFp = (VSILFILE*) VSICreateCachedFile ( (VSIVirtualHandle*) m_poFp);
+        }
+        VSIFSeekL(m_poFp, sizeof(magicbytes), SEEK_SET); // skip magic bytes
+        uoffset_t headerSize;
+        VSIFReadL(&headerSize, sizeof(uoffset_t), 1, m_poFp);
+        auto featuresCount = m_poHeader->features_count();
+        auto treeSize = PackedRTree::size(featuresCount);
+        OGREnvelope env;
+        m_poFilterGeom->getEnvelope(&env);
+        Rect r { env.MinX, env.MinY, env.MaxX, env.MaxY };
+        CPLDebug("FlatGeobuf", "Spatial index search on %f,%f,%f,%f", env.MinX, env.MinY, env.MaxX, env.MaxY);
+        auto readNode = [this, headerSize] (uint8_t *buf, uint32_t i, uint32_t s) {
+            VSIFSeekL(m_poFp, sizeof(magicbytes) + sizeof(uoffset_t) + headerSize + i, SEEK_SET);
+            VSIFReadL(buf, 1, s, m_poFp);
+        };
+        m_foundFeatureIndices = PackedRTree::streamSearch(featuresCount, 16, r, readNode);
+        m_featuresCount = m_foundFeatureIndices.size();
+        if (m_featuresCount == 0) {
+            CPLDebug("FlatGeobuf", "No found features in spatial index search");
+            return;
+        }
+        CPLDebug("FlatGeobuf", "%zu features found in spatial index search", m_featuresCount);
+        m_featureOffsets = static_cast<uint64_t *>(VSI_MALLOC_VERBOSE(featuresCount * 8));
+        VSIFSeekL(m_poFp, sizeof(magicbytes) + sizeof(uoffset_t) + headerSize + treeSize, SEEK_SET);
+        VSIFReadL(m_featureOffsets, 8, featuresCount, m_poFp);
+    } else {
+        CPLDebug("FlatGeobuf", "processSpatialIndex noop");
+    }
+}
+
+GIntBig OGRFlatGeobufLayer::GetFeatureCount(int /* bForce */) {
+    processSpatialIndex();
```

Comment:
the attribute filter should also be taken into account (by possibly using the base/slow implementation)

Reply:
you can check the 'compliance' of the driver with the test_ogrsf utility (cd apps; make test_ogrsf)

Reply:
Very useful! Was able to fix many issues after using test_ogrsf but I'm getting some unexplained errors related to GetFeature and filtering so I'll have to leave this open and revisit.